mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java

```diff
@@ -622,20 +622,43 @@ private void assertIsOwner(MailboxSession mailboxSession, MailboxPath mailboxPat
     }
 
     @Override
-    public Flux<MailboxMetaData> search(MailboxQuery expression, MailboxSession session) {
+    public Flux<MailboxMetaData> search(MailboxQuery expression, MailboxSearchFetchType fetchType, MailboxSession session) {
         Mono<List<Mailbox>> mailboxesMono = searchMailboxes(expression, session, Right.Lookup).collectList();
-        MessageMapper messageMapper = mailboxSessionMapperFactory.getMessageMapper(session);
 
         return mailboxesMono
-            .flatMapMany(mailboxes -> Flux.fromIterable(mailboxes)
-                .filter(expression::matches)
-                .flatMap(mailbox -> retrieveCounters(messageMapper, mailbox, session)
-                    .map(Throwing.<MailboxCounters, MailboxMetaData>function(
-                        counters -> toMailboxMetadata(session, mailboxes, mailbox, counters))
-                        .sneakyThrow())))
+            .flatMapMany(mailboxes -> {
+                if (fetchType == MailboxSearchFetchType.Counters) {
+                    return withCounters(expression, session, mailboxes);
+                } else {
+                    return withoutCounters(expression, session, mailboxes);
+                }
+            })
```

Comment:
How about bring the if block outside so we only need to run the check once?

Reply:
The if branching is cheap, but ok.